### PR TITLE
Transport client: Don't validate node in handshake

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #30141. Right now in the transport client we open a
temporary node connection and take the node information. This node
information is used to open a permanent connection that is used for the
client. However, we continue to use the configured transport address.
If the configured transport address is a load balancer, you might
connect to a different node for the permanent connection. This causes
the handshake validation to fail. This commit removes the handshake
validation for the transport client.

### CODE REVIEW

**Comment by tbrooks8** [comment]
I want to point out that there is some tension with this functionality and how we publicize (and log) discovery node information. You might not actually be connected to the node that you think you are connected to (https://github.com/elastic/elasticsearch/issues/30141#issuecomment-388137656). 

It feels kind of weird to me that the transport client is based on "nodes" and not "connections". But that ship has sailed and is probably not too important to resolve for something that is deprecated. But I want to raise that point. Should I modify the PR to store the node that is eventually connected to in the "persistent, managed" connection?

File: server/src/main/java/org/elasticsearch/transport/TransportService.java [hunk 1]
```diff
@@ -325,13 +330,25 @@ public void connectToNode(DiscoveryNode node) throws ConnectTransportException {
...
+
+    /**
+     * Connect to the specified node with the given connection profile. This method allows the caller to
+     * specify whether the node connection should be validated against the specified node.
+     *
+     * @param node the node to connect to
+     * @param connectionProfile the connection profile to use when connecting to this node
+     * @param validateNodeConnection boolean indicating if the node connection should be validated
+     */
+    public void connectToNode(final DiscoveryNode node, ConnectionProfile connectionProfile, boolean validateNodeConnection) {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/TransportService.java]: can we please not add another API here but rather try to detect if we are a client or a node? I mean down the road we don't want to support this anymore so lets make sure we don't have special public APIs. I can imagine a boolean on the ctor or even check if ClusterService is != null (not that intuitive). You can just add an overloaded ctor instead.

**Review by s1monw — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/transport/TransportService.java (hunk 1 above)
Comment (s1monw) [reply server/src/main/java/org/elasticsearch/transport/TransportService.java]: we can also just look at the settings like this `TransportClient.CLIENT_TYPE.equals(settings.get(Client.CLIENT_TYPE_SETTING_S.getKey()));`

**Review by s1monw — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java [hunk 2]
```diff
@@ -361,10 +363,10 @@ public void sample() {
         protected abstract void doSample();
 
         /**
-         * validates a set of potentially newly discovered nodes and returns an immutable
-         * list of the nodes that has passed.
+         * Establishes the node connections. If validateInHandshake is set to true, the connection will fail if
+         * node returned in the handshake response is different than the discovery node.
          */
-        protected List<DiscoveryNode> validateNewNodes(Set<DiscoveryNode> nodes) {
+        List<DiscoveryNode> establishNodeConnections(Set<DiscoveryNode> nodes, boolean validateInHandshake) {
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java]: `validateInHandshake` not used anymore?

**Review by bleskes — APPROVED** [review]
LGTM. Thanks @tbrooks8

**Pull request merged**